api/application.go

```diff
@@ -296,6 +304,103 @@ func (h ApplicationHandler) BucketDelete(ctx *gin.Context) {
 	h.delete(ctx, &m.BucketOwner)
 }
 
+// TagList godoc
+// @summary List tag references.
+// @description List tag references.
+// @tags get
+// @produce json
+// @success 200 {object} []api.Ref
+// @router /applications/{id}/tags/id [get]
+// @param id path string true "Application ID"
+func (h ApplicationHandler) TagList(ctx *gin.Context) {
+	id := h.pk(ctx)
+	p := &model.Application{}
+	result := h.DB.First(p, id)
+	if result.Error != nil {
+		h.getFailed(ctx, result.Error)
+		return
+	}
+	db := h.DB.Model(p).Association("Tags")
+	list := []model.Tag{}
+	err := db.Find(&list)
+	if err != nil {
+		h.listFailed(ctx, err)
+		return
+	}
+	resources := []Ref{}
+	for i := range list {
+		r := Ref{}
+		r.With(list[i].ID, list[i].Name)
+		resources = append(resources, r)
+	}
+	ctx.JSON(http.StatusOK, resources)
+}
+
+// TagAdd godoc
+// @summary Add tag association.
+// @description Ensure tag is associated with the application.
+// @tags create
+// @accept json
+// @produce json
+// @success 201 {object} api.Ref
+// @router /tags [post]
+// @param tag body Ref true "Tag data"
+func (h ApplicationHandler) TagAdd(ctx *gin.Context) {
+	id := h.pk(ctx)
+	r := &Ref{}
+	err := ctx.BindJSON(r)
+	if err != nil {
+		h.bindFailed(ctx, err)
+		return
+	}
+	p := &model.Application{}
+	result := h.DB.First(p, id)
+	if result.Error != nil {
+		h.getFailed(ctx, result.Error)
+		return
+	}
+	db := h.DB.Model(p).Association("Tags")
+	m := &model.Tag{}
+	m.ID = r.ID
+	err = db.Append(m)
+	if err != nil {
+		h.createFailed(ctx, err)
+		return
+	}
+	ctx.JSON(http.StatusCreated, r)
+}
+
+// TagDelete godoc
+// @summary Delete tag association.
+// @description Ensure tag is not associated with the application.
+// @tags delete
+// @success 204
+// @router /applications/{id}/tags/{sid} [delete]
+// @param id path string true "Application ID"
+// @param sid path string true "Tag ID"
+func (h ApplicationHandler) TagDelete(ctx *gin.Context) {
+	id := h.pk(ctx)
+	sid := ctx.Param(ID2)
+	p := &model.Application{}
+	result := h.DB.First(p, id)
+	if result.Error != nil {
+		h.getFailed(ctx, result.Error)
+		return
+	}
+	db := h.DB.Model(p).Association("Tags")
+	m := &model.Tag{}
+	n, _ := strconv.Atoi(sid)
+	id = uint(n)
```

Comment:
Personally I find little bit confusing the variables naming in this method (id, sid, id2, m, n), but only thing I'd ask is about the ```id``` variable assignment on this line, there doesn't seem to be any use of it (unless I'm missing something).

Reply:
I will clean this up.

Reply:
Fixed.